### PR TITLE
8303920: Avoid calling out to python in DataDescriptorSignatureMissing test

### PR DESCRIPTION
**Pull request by eirbjo**
Please review this PR which brings  the DataDescriptorSignatureMissing test back to life.

This test currently calls out to Python to create a test vector ZIP with a Data Descriptor without the recommended but optional signature. The Python dependency has turned out to be very brittle, so the test is currently marked with `@ignore` 

The PR replaces Python callouts with directly creating the test vector ZIP in the test itself. We can then remove the `@ignore`tag and run this useful test automatically.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303920](https://bugs.openjdk.org/browse/JDK-8303920): Avoid calling out to python in DataDescriptorSignatureMissing test (**Enhancement** - P4)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**)

### Contributors
 * Jaikiran Pai `<jpai@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12959/head:pull/12959` \
`$ git checkout pull/12959`

Update a local copy of the PR: \
`$ git checkout pull/12959` \
`$ git pull https://git.openjdk.org/jdk.git pull/12959/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12959`

View PR using the GUI difftool: \
`$ git pr show -t 12959`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12959.diff">https://git.openjdk.org/jdk/pull/12959.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12959#issuecomment-1462701573)

### CODE REVIEW

File: test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
Comment (jaikiran) [inline test/jdk/java/util/zip/DataDescriptorSignatureMissing.java:29]: Hello Eirik, I think the summary should no longer have references to python. Since this test was introduced for https://bugs.openjdk.org/browse/JDK-8056934, perhaps we can just change the `@summary` of this test definition to say:
>
>@summary Verify the ability to read zip files whose local header data descriptor is missing the optional signature

Comment (eirbjo) [reply test/jdk/java/util/zip/DataDescriptorSignatureMissing.java]: Thanks, I updated the summary according to your suggestion. I left it there to keep the history/context of the test, but I guess that's a bit detailed and is already captured in the bug.

File: test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
Comment (eirbjo) [reply test/jdk/java/util/zip/DataDescriptorSignatureMissing.java]: I have no affiliation to Google. I reverted the change to their copyright line and added an Oracle one.

There is very little  code left from Martin here  (the Zip64 note in the header, the class name and some whitespace), but let's keep it anyhow.

**Comment by jaikiran** [comment]
It's good that this is being revived and no longer relies on an external tool/program to generate a zip which was triggering the issue in https://bugs.openjdk.org/browse/JDK-8056934.

The changes look good to me. In order to verify that without the fix in https://bugs.openjdk.org/browse/JDK-8056934, this test fails (i.e. reproduces the issue), I reverted the fix that was done in that issue (`git revert 3951dda4cf06c6e61e19d3df26a792022c1701b9`) and then built the JDK and ran this updated test. It's run into a NullPointerException within the test because the second entry in the zip is missing after the zip is read through the `ZipInputStream`. Could you add this following patch to your test:

```diff
diff --git a/test/jdk/java/util/zip/DataDescriptorSignatureMissing.java b/test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
index 5efdc59de63..636cecb4851 100644
--- a/test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
+++ b/test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
@@ -39,6 +39,7 @@ import java.nio.charset.StandardCharsets;
 import java.util.zip.*;
 
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
 
 public class DataDescriptorSignatureMissing {
 
@@ -55,10 +56,12 @@ public class DataDescriptorSignatureMissing {
         try (ZipInputStream in = new ZipInputStream(
                 new ByteArrayInputStream(zip))) {
             ZipEntry first = in.getNextEntry();
+            assertNotNull(first, "Zip file is unexpectedly missing first entry");
             assertEquals(first.getName(), "first");
             assertEquals(in.readAllBytes(), "first".getBytes(StandardCharsets.UTF_8));
 
             ZipEntry second = in.getNextEntry();
+            assertNotNull(second, "Zip file is unexpectedly missing second entry");
             assertEquals(second.getName(), "second");
             assertEquals(in.readAllBytes(), "second".getBytes(StandardCharsets.UTF_8));
         }

```

so that instead of running into a NullPointerException, the test will (rightly) reproduce and report the missing second entry?

**Comment by eirbjo** [comment]
Nice! I added your assertNotNulls and tested with the off-by-one code which gives me the following:

```
test DataDescriptorSignatureMissing.shouldParseSignaturelessDescriptor(): failure
java.lang.AssertionError: Zip file is unexpectedly missing second entry expected object to not be null
```

**Comment by eirbjo** [comment]
> This now looks good to me.

Thanks taking time to do this thorough review and especially for running the regressing case. Much appreciated!

I'll wait for a second review before integrating.

**Comment by eirbjo** [comment]
Since we strip 4 bytes from the first entry's data descriptor, we need to account for this by reducing the second CEN header's LOC offset by 4. Similarly, the END header's CEN offset also needs adjustment.

**Comment by eirbjo** [comment]
Reopening this PR.

Before being closed for inactivity, this PR was reviewed by @jaikiran, who requested that another reviewer with experience in this area also take a look at it before integration.

I think it would be good to have this regression test run automatic again.

**Review by LanceAndersen — COMMENTED** [review]
Thanks for trying to move this forward.  Please see my comments below.

Another question, is the zip that is generated by this test readable by other zip tools such as info-zip, Apache Common-compress, winzip?

**Comment by eirbjo** [comment]
> Another question, is the zip that is generated by this test readable by other zip tools such as info-zip, Apache Common-compress, winzip?

- info-zip: Does not support unzipping from a zip,  so uses the CEN instead of the data descriptor.
- Apache commons-compress, reads the signature-less ZIP just fine.
- winzip: I do not currently have easy access to Windows, so can't test this. But I would assume it also uses the CEN when unzipping

File: test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
Comment (eirbjo) [reply test/jdk/java/util/zip/DataDescriptorSignatureMissing.java:3]: This way of updating the copyright was suggested by @jaikiran in the March 10th comment above. Would be nice to get this clarified, yes.

File: test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
Comment (eirbjo) [reply test/jdk/java/util/zip/DataDescriptorSignatureMissing.java:109]: Added a comment including some structural examples. (I personally feel it maybe ended up a bit excessive)

File: test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
Comment (eirbjo) [reply test/jdk/java/util/zip/DataDescriptorSignatureMissing.java:3]: There is actually very little left of Martin's code after my rewrite, besides whitespace, some curly braces and a couple of imports. Hardly defensible IP, but then I Am Not a Lawyer :-)

**Comment by LanceAndersen** [comment]
Thinking some more about this, I would like to see us keep the Zip generated by python, store it in a byte array (or equivalent) as it also validate that we can still process the zip given this was the original test and the zip is being generated by a 3rd party tool.

This would be an additional test along with your proposed enhancements

**Comment by eirbjo** [comment]
> Thinking some more about this, I would like to see us keep the Zip generated by python, store it in a byte array (or equivalent) as it also validate that we can still process the zip given this was the original test and the zip is being generated by a 3rd party tool.

Lance, 

I was finally able to reproduce the original issue using Python 3.4.4. (Which was a challenge to install given its archaic dependencies!)

I was able to verfy that the missing signature is the ONLY difference between the input and output files. (Except updated LOC and CEN offsets accounting for the missing bytes). Additionally, I independently removed the signature files from the input file, this produced an output file binary identical to Python's.

Given that the one and only difference introduced by the Python script is covered by the test in this PR, I'm not sure I see any additional value in adding a test with the binary test vector produced by Python. I think it will just increase our maintenance costs, without adding any real value or coverage.  

If you see this differently, that's of course ok. Just let me know and I'll create the test with the encoded binary ZIP (which I have easily available now).

Waiting for your guidance, thanks :-)

**Comment by LanceAndersen** [comment]
> > Thinking some more about this, I would like to see us keep the Zip generated by python, store it in a byte array (or equivalent) as it also validate that we can still process the zip given this was the original test and the zip is being generated by a 3rd party tool.
> 
> Lance,
> 
> I was finally able to reproduce the original issue using Python 3.4.4. (Which was a challenge to install given its archaic dependencies!)
> 
> I was able to verfy that the missing signature is the ONLY difference between the input and output files. (Except updated LOC and CEN offsets accounting for the missing bytes). Additionally, I independently removed the signature files from the input file, this produced an output file binary identical to Python's.
> 
> Given that the one and only difference introduced by the Python script is covered by the test in this PR, I'm not sure I see any additional value in adding a test with the binary test vector produced by Python. I think it will just increase our maintenance costs, without adding any real value or coverage.
> 
> If you see this differently, that's of course ok. Just let me know and I'll create the test with the encoded binary ZIP (which I have easily available now).
> 
> Waiting for your guidance, thanks :-)

If you verified that we are a complete match, I am good with that.  Thank you for validating

File: test/jdk/java/util/zip/DataDescriptorSignatureMissing.java
Comment (LanceAndersen) [reply test/jdk/java/util/zip/DataDescriptorSignatureMissing.java:3]: I exchanged messages with Iris and she is comfortable with the updates to the copyright

**Review by LanceAndersen — APPROVED** [review]
I think you are in a good place with the test clean up.  Thanks for your efforts here

I can kick of a test run internally next week or perhaps Sunday

**Review by irisclark — APPROVED** [review]

**Comment by eirbjo** [comment]
> I can kick of a test run internally next week or perhaps Sunday

Thanks for your reviews, Lance and Iris!

FWIW, the test ran fine on Github Actions, including on `linux-x86` (which is 32-bit, right?):

```
TEST: java/util/zip/DataDescriptorSignatureMissing.java
  build: 0.081 seconds
  compile: 0.081 seconds
  junit: 0.055 seconds
TEST RESULT: Passed. Execution successful
```

https://github.com/eirbjo/jdk/actions/runs/6696663322/job/18196803138

**Comment by eirbjo** [comment]
> > I can kick of a test run internally next week or perhaps Sunday
> > 
> FWIW, the test ran fine on Github Actions, including on `linux-x86` (which is 32-bit, right?):

Disregard this comment, since it confuses this PR with #12991 :-)

**Comment by LanceAndersen** [comment]
> > I can kick of a test run internally next week or perhaps Sunday
> 
> Thanks for your reviews, Lance and Iris!
> 
> FWIW, the test ran fine on Github Actions, including on `linux-x86` (which is 32-bit, right?):
> 
> ```
> TEST: java/util/zip/DataDescriptorSignatureMissing.java
>   build: 0.081 seconds
>   compile: 0.081 seconds
>   junit: 0.055 seconds
> TEST RESULT: Passed. Execution successful
> ```
> 
> https://github.com/eirbjo/jdk/actions/runs/6696663322/job/18196803138

Have an internal mach5 run going and will let you know when it completes

**Review by LanceAndersen — APPROVED** [review]
Latest Mach 5 run looks fine

**Review by jaikiran — APPROVED** [review]
Hello Eirik, thank you for this change - it looks really good.

**Comment by jaikiran** [comment]
Lance and Iris too have approved the PR and Lance has noted that the mach5 run came back fine. I'll go ahead and sponsor this now.

/sponsor

**Pull request closed without merging**
